File: Java/src/main/java/testingwindows/LateData.java
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package testingwindows;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.coders.StringUtf8Coder;
import org.apache.beam.sdk.options.PipelineOptions;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.testing.TestStream;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.GroupByKey;
import org.apache.beam.sdk.transforms.ParDo;
import org.apache.beam.sdk.transforms.windowing.FixedWindows;
import org.apache.beam.sdk.transforms.windowing.Window;
import org.apache.beam.sdk.values.KV;
import org.apache.beam.sdk.values.TimestampedValue;
import org.joda.time.Duration;
import org.joda.time.Instant;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class LateData {
    private static final Logger LOG = LoggerFactory.getLogger(LateData.class);

    public static void main(String[] args) {

        PipelineOptions options = PipelineOptionsFactory.fromArgs(args).withValidation().as(PipelineOptions.class);

        Pipeline p = Pipeline.create(options);

        Integer windowLength = 10;
        Integer allowLateSize = 3;

        TestStream<String> streamEvents = TestStream.create(StringUtf8Coder.of())
                .addElements(
                        TimestampedValue.of("A", new Instant(0)),
                        TimestampedValue.of("B", new Instant(4000)), // Value of Instant is in millis
                        TimestampedValue.of("C", new Instant(6000)),
                        TimestampedValue.of("D", new Instant(6000))
                )
                // Move the watermark past the end the end of the window
                .advanceWatermarkTo(new Instant(9000))
                .advanceProcessingTime(Duration.standardSeconds(9))
                .addElements(
                        // element arrives at second 9 but with timestamp 0
                        TimestampedValue.of("Late-but-before-window-closing", new Instant(0))
                )
                .advanceWatermarkTo(new Instant(11000))// move Watermark time to sec 11
                .advanceProcessingTime(Duration.standardSeconds(2)) // move Processing time to sec 11
                .addElements(
                        // element arrives at sec 11 but with ts 8, outside the window but within allowed lateness
                        TimestampedValue.of("Late-but-allowed", new Instant(9000))
                )
                .advanceWatermarkTo(new Instant(15000))
                .advanceProcessingTime(Duration.standardSeconds(4)) // move Processing time to sec 15
                .addElements(
                        // element arrives at sec 15 but with ts 9, outside the window
                        TimestampedValue.of("Late-outside-window-and-latedata", new Instant(9000)),
                        TimestampedValue.of("window2", new Instant(14000))

                )
                .advanceWatermarkToInfinity();

        p.apply(streamEvents)
                .apply("KVs", ParDo.of(new DoFn<String, KV<String, KV<String, Instant>>>() {
                    @ProcessElement
                    public void processElement(ProcessContext c) {
                        KV<String, Instant> element = KV.of(c.element(), c.timestamp());
                        c.output(KV.of("key", element));
                    }
                }))
                .apply(Window.<KV<String, KV<String, Instant>>>into(
                        FixedWindows.of(Duration.standardSeconds(windowLength)))
                        .withAllowedLateness(Duration.standardSeconds(allowLateSize))
                        .discardingFiredPanes()
                )
                .apply(GroupByKey.create())
                .apply("Log", ParDo.of(new DoFn<KV<String, Iterable<KV<String, Instant>>>, String>() {
                    @ProcessElement
                    public void processElement(ProcessContext c) {
                        LOG.info("TRIGGER \n" + c.element().getValue().toString()
                                + ". \n Window end " + c.timestamp().plus(1));
                        c.output(c.pane().toString());
                    }
                }));

        p.run();
    }
}

        /*
        EXPLANATION
        1 - Elements A,B,C,D arrive to the pipeline on time
        1 - Element "Late-but-before-window-closing" arrives 9s late, but before window closes
        2 - Element "Late-but-allowed" arrives after window closing, but within the allowed late time
        3 - Element "Late-outside-window-and-latedata" arrives after window closing and after allowed late time
        4 - Element "window-2" arrives on time but in a different window than 0 to 10000

        The first trigger with elements (1) contains all data from the first window that arrived before window closed
        The second trigger with element (2) includes the data that arrived late but within the allowed late time
        Elements in (3) are discarded, since they arrive after the allowed late time
        The third trigger with element (4) appears in a new windows that closes at 1970-01-01T00:00:20.000Z
         */

File: Python/testing_windows/late_data.py
#  Copyright 2022 Google LLC
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

# standard libraries
import logging

# third party libraries
from apache_beam import DoFn, GroupByKey, Map, ParDo, WindowInto
from apache_beam.options.pipeline_options import PipelineOptions
from apache_beam.testing.test_pipeline import TestPipeline
from apache_beam.testing.test_stream import TestStream, TimestampedValue
from apache_beam.transforms.trigger import AccumulationMode
from apache_beam.transforms.window import FixedWindows


def get_input_stream():
    stream = (
        TestStream()
        .add_elements(
            [
                # timestamp value in seconds since epoch
                TimestampedValue("A", timestamp=0),
                TimestampedValue("B", timestamp=4),
                TimestampedValue("C", timestamp=6),
                TimestampedValue("D", timestamp=6),
            ]
        )
        .advance_watermark_to(9)
        .advance_processing_time(9)
        .add_elements(
            [TimestampedValue("Late-but-before-window-closing", timestamp=0)]
        )
        .advance_watermark_to(11)
        .advance_processing_time(3)
        .add_elements(
            [
                # Although pane fired, below element still within lateness configured
                TimestampedValue("Late-but-allowed", timestamp=9)
            ]
        )
        .advance_watermark_to(15)
        .advance_processing_time(4)
        .add_elements(
            [
                TimestampedValue("Late-and-outside-window", timestamp=9),
                TimestampedValue("Window2-E", timestamp=11),
            ]
        )
        .advance_watermark_to_infinity()
    )
    return stream


class FormatElements(DoFn):
    # Formats the elements and adds metadata information to the pane
    def process(
        self, element, window=DoFn.WindowParam, pane=DoFn.PaneInfoParam
    ):
        yield "Elements in pane: {} \nWindow {} - Pane Index {}\n".format(
            element[1], window, pane.index
        )


def run():
    options = PipelineOptions(streaming=True)

    p = TestPipeline(options=options)

    window_size_seconds = 10
    window_allowed_lateness_seconds = 3

    stream = get_input_stream()

    (
        p
        | stream
        | "Assign elements within pane same key" >> Map(lambda e: ("Key", e))
        | "Window into Discarding FixedWindows"
        >> WindowInto(
            FixedWindows(size=window_size_seconds),
            allowed_lateness=window_allowed_lateness_seconds,
            # Set `accumulation_mode` as `ACCUMULATING` when you want to collect all the
            # panes fired in the window. To have the panes be considered individually
            # use `DISCARDING` mode. Use `DISCARDING` when the datum are indipendent
            # of each other. `ACCUMULATING` will fire each time the a panes arrives from a window
            accumulation_mode=AccumulationMode.DISCARDING,
        )
        | "Group together elements within same pane and key" >> GroupByKey()
        | "Format elements" >> ParDo(FormatElements())
        | Map(print)
    )

    p.run()


if __name__ == "__main__":
    logging.getLogger("LateData").setLevel(logging.INFO)
    run()


#   EXPLANATION
# 1 - Elements A,B,C,D arrive to the pipeline on time
# 1 - Element "Late-but-before-window-closing" arrives 9s late, but before window closes
# 2 - Element "Late-but-allowed" arrives after window closing, but within the allowed late time
# 3 - Element "Late-and-outside-window" arrives after window closing and after allowed late time
# 4 - Element "window-2" arrives on time but in a different window than 0 to 10000
#
# The first trigger with elements (1) contains all data from the first window that arrived before window closed
# The second trigger with element (2) includes the data that arrived late but within the allowed late time
# Elements in (3) are discarded, since they arrive after the allowed late time
# The third trigger with element (4) appears in a new window `[10.0, 20.0)` with pane index `0`
# Even though the Window is same for (1) (2) the pane index for (1) is `0` and for (2) is `1`
# denoting they are logically separated
